### DISCUSSION TITLE
Protocol inheritance with property

### DISCUSSION THREAD

**Discussion by ZashIn** [microsoft/pyright#6203] [Q&A]
How do you define a Protocol that requires an attribute of any kind (property, class var, instance var, ...) to be **at least** readable, i.e. `obj.id: int`?
This was already described in #2371 in [this comment](https://github.com/microsoft/pyright/issues/2371#issuecomment-934606895), but does not work with inheritance:
```py
from typing import Protocol

class Identifiable(Protocol):
    # How type `obj.id: int` to be at least readable?
    @property
    def id(self) -> int: ...

class One(Identifiable):
    def __init__(self, id: int = 1) -> None:
        self.__id = id

    @property
    def id(self) -> int:
        return self.__id

class Two(Identifiable):  # Cannot use inheritance here
    id: int = 2  # "id" incorrectly overrides property of same name in class "Identifiable" (reportIncompatibleMethodOverride)

class Three:
    id: int = 3

def test(o: Identifiable):
    print(o.id)

# All ok
test(One())
test(Two())
test(Three())
```
mypy does not complain about this ([PR](https://github.com/python/mypy/pull/14081)).

Any alternative with pyright?

**Comment by erictraut** [comment]
The protocol definition in your code sample looks fine to me. [PEP 544](https://peps.python.org/pep-0544/#protocol-members) says:

> To define a read-only protocol variable, one can use an (abstract) property.

Pyright supports this for purposes of protocol matching, as you can see with `test(Three())`.

For direct inheritance checks, pyright applies a stricter test. If a base class defines a property, a subclass must also define a property. This is important because an access to a property and a class variable vary when accessing the attribute through the class object (as opposed to an instance of the class).

```python
class A:
    @property
    def id(self) -> int:
        return 1

class B:
    id: int = 1

print(A.id) # <property object>
print(B.id) # 1
```

Mypy is more lax when it comes to tracking instance variables versus class variables. (There are many open issues in its issue tracker about this.)

There has been some discussion about adding a new `ReadOnly` special form that could be used to better handle this case.

**Reply by ZashIn** [reply]
I suppose it depends on how to interpret a Protocol, but I understand the reasoning.

A clear distinction between a `ClassVar` and an instance war would help, too.
Something like the this:
```py
class A(Protocol):
    id: ReadOnly[InstVar[int]]

class B(Protocol):
    id: InstVar[int]

    # or a @InstVar here
    @property
    def id(self) -> int:  ...
```
